### ISSUE TITLE
Need AJAX supported - Auth module 

### ISSUE THREAD

**Issue by dannymk** [AlliumCepa/webgui#41]
We can probably subclass WebGUI::Auth::WebGUI to provide login capabilities via AJAX requests.  The following will probably work out of the box:

```
sub www_login {
   my $self = shift;
   if(!$self->authenticate($self->session->form->process("username"),$self->session->form->process("identifier"))){
      $self->session->response->status(401);
```

However, all responses have to use JSON.

An authenticated response should probably be in the form:

http-status:  200

```
{
     "data" : [
           "identity" : {
                 "first" : "Daniel",
                 "last" : "MyLastName",
                 "avatar" : "https://myavatar.com/daniel.jpg"
           },
           "joined" : "2019-04-01",
           "session" : "6c9676d699-2nj8w-7bc874d4bd-bt2kk"
     ]
}
```

An invalid response should be something like:
http-status: 4xx
{
     "code" : "4xx",
     "message" : "...$i18n->get(70)"
}

See:  https://www.restapitutorial.com/httpstatuscodes.html

**Comment by dannymk** [comment]
We should be able to do something like $self->session->json->to_json($output);

**Comment by scrottie** [comment]
How should we handle the case when wG tries to force the user to update their expired password?

Easy option would be to treat it as an error with a message that they need to return to the site and log in there.

**Comment by scrottie** [comment]
More notes in the commit messages (now easily linked from here... yay... I swear, I was doing that all of the time for $daycontract but just brainfarted on it here).

JSON responses are triggered by specifying `application/json` in the `Accept` header of the request:

```
 curl 'http://127.0.0.1:5000/home' -H 'Accept: application/json' -H 'Content-Type: application/x-www-form-urlencoded' -H 'Cookie: wgSession=3wnJeyDlRXanBqChk5gH1A' --data 'op=auth&method=login&username=admin&identifier=123qwe'

{"data":["identity",{"avatar":"","last":null,"first":null},"joined","2002-04-27","session","3wnJeyDlRXanBqChk5gH1A"]}
```

When existing REST APIs are perfect for what you need, like login (okay, since it has op and method params instead of path, it's not exactly REST, but we could do an /op/auth/method/login thing if we wanted), I'll make it look at the requested datatype and respond accordingly.  When we need new API endpoints, I'll assume JSON.

**Comment by dannymk** [comment]
The returned data does not match the spec.  Furthermore it is not valid JSON as there are keys and values and the current return only has keys similar to a list.

**Comment by scrottie** [comment]
Haha, it does match the spec, and that's the problem.

json_pp doesn't barf on it.  There are probably better validation tools but it at least thinks it is valid.

I'll fix it.  One sec.

**Comment by scrottie** [comment]
After pretty printing, it looks like this now:

```
{
   "data" : {
      "identity" : {
         "avatar" : "",
         "first" : null,
         "last" : null
      },
      "joined" : "2002-04-27",
      "session" : "3wnJeyDlRXanBqChk5gH1A"
   }
}
```

**Comment by dannymk** [comment]
Ah, I was just fixing that too :-)  You are right, the spec was wrong.

**Comment by dannymk** [comment]
Available in #55ec3baf8b2ee85c17b6d54d08f82599a615a347
